File: 547-number-of-provinces/number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:

        n = len(isConnected)
        adj_list = defaultdict(list)

        for i in range(n):
            for j in range(n):
                if isConnected[i][j] == 1:
                    adj_list[i].append(j)

        # print("Adjacency list : ", adj_list)
        visited = set()
        count = 0 

        def dfs(node):
            visited.add(node)

            for neighbor in adj_list[node]:
                if neighbor not in visited:
                    dfs(neighbor)

        for node in range(n):
            if node not in visited:
                dfs(node)
                count += 1

        print("count : ", count)
        return count

File: 547-number-of-provinces/number-of-provinces.java
class Solution {
    public void dfs(int i, boolean[] visited, List<List<Integer>> adjList) {
        visited[i] = true;
        for (int n : adjList.get(i)){
            if (!visited[n]) {
                dfs(n, visited, adjList);
            }
        }
    };

    public void bfs(int i, boolean[] visited, List<List<Integer>> adjList) {

        Queue<Integer> q = new LinkedList<>();
        q.offer(i);
        visited[i] = true;

        while (!q.isEmpty()) {
            int node = q.poll();

            for (int n : adjList.get(node)){
                if (!visited[n]){
                    q.offer(n);
                    visited[n] = true;

                }
            }
        }
    }

    public int findCircleNum(int[][] isConnected) {
        List<List<Integer>> adjList = new ArrayList<>();
        int n = isConnected.length;

        for (int i = 0; i < n; i++) {
            adjList.add(new ArrayList<>());
            for (int j = 0; j < n; j++){
                if (i != j && isConnected[i][j] == 1){
                    adjList.get(i).add(j);
                }
            }
        }
        int count = 0;
        boolean[] visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                bfs(i, visited, adjList);
                count++;
            }
        }
        return count;
    }
}